### PR TITLE
Changes for working on a quadcopter

### PR DESCRIPTION
**Pull request by raabuchanan**
Changes:
-Adds Stereo rectification
-Adds camera to body transform
-Makes visualization a ROS param
-Changes rostopic names

### CODE REVIEW

File: include/orb_slam_2_ros/interface_stereo.hpp [hunk 1]
```diff
@@ -36,6 +42,11 @@ class OrbSlam2InterfaceStereo : public OrbSlam2Interface {
   std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>> left_sub_;
   std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>> right_sub_;
   std::shared_ptr<message_filters::Synchronizer<sync_pol>> sync_;
+
+  cv::Mat M1l_,M2l_,M1r_,M2r_;
```
Comment (alexmillane) [inline include/orb_slam_2_ros/interface_stereo.hpp]: Spacing. What are these matrices? Please leave a comment. If they are transforms name as T_A_B_.

File: include/orb_slam_2_ros/interface.hpp [hunk 2]
```diff
@@ -27,6 +29,9 @@ class OrbSlam2Interface {
   OrbSlam2Interface(const ros::NodeHandle& nh,
                     const ros::NodeHandle& nh_private);
 
+  virtual bool imagePreProcessing()
```
Comment (alexmillane) [inline include/orb_slam_2_ros/interface.hpp]: Purpose of this function?

File: include/orb_slam_2_ros/interface.hpp [hunk 3]
```diff
@@ -43,6 +48,9 @@ class OrbSlam2Interface {
   // Helper functions
   void convertOrbSlamPoseToKindr(const cv::Mat& T_cv, Transformation* T_kindr);
 
+  virtual bool getBodyTransform()
```
Comment (alexmillane) [inline include/orb_slam_2_ros/interface.hpp]: As above.

In addition please autoformat your code with clang format. If you use sublime, this is available here. A similar package exists for eclipse.

https://packagecontrol.io/packages/Clang%20Format

File: src/library/interface.cpp [hunk 4]
```diff
@@ -14,16 +14,18 @@ OrbSlam2Interface::OrbSlam2Interface(const ros::NodeHandle& nh,
...
+
   advertiseTopics();
   getParametersFromRos();
 }
 
 void OrbSlam2Interface::advertiseTopics() {
   // Advertising topics
   T_pub_ = nh_private_.advertise<geometry_msgs::TransformStamped>(
-      "transform_cam", 1);
+      "orb_slam_2", 1);
```
Comment (alexmillane) [inline src/library/interface.cpp]: Why this change? The topic is publishing in the node's private namespace (for example orb_slam_2), this change will lead to a redundant and uninformative change.

File: src/library/interface_mono.cpp [hunk 5]
```diff
@@ -9,20 +9,46 @@ OrbSlam2InterfaceMono::OrbSlam2InterfaceMono(const ros::NodeHandle& nh,
     : OrbSlam2Interface(nh, nh_private) {
   // Getting data and params
   subscribeToTopics();
-  //advertiseTopics();
-  //getParametersFromRos();
+
+  if(use_body_transform_){
+    if(!getBodyTransform()){
+      ros::shutdown();
+      exit(1);
```
Comment (alexmillane) [inline src/library/interface_mono.cpp]: I think I see the logic here... However, I would rather check for the body transform when loading parameters from the ROS param server (see below).

File: src/library/interface_mono.cpp [hunk 6]
```diff
@@ -9,20 +9,46 @@ OrbSlam2InterfaceMono::OrbSlam2InterfaceMono(const ros::NodeHandle& nh,
...
+
+  cv::Mat T_C0_B;
+  Transformation T_C_B;
+
+  fsSettings["T_c0_fcuimu"] >> T_C0_B;
+  convertOrbSlamPoseToKindr(T_C0_B, &T_C_B);
+  T_B_C_ = T_C_B.inverse();
+
+  return true;
+}
```
Comment (alexmillane) [inline src/library/interface_mono.cpp]: There are a few ways to load parameters. Our policy is to try to use the ROS parameter server exclusively. A good example of how to do this can be found here:

https://github.com/ethz-asl/voxblox/blob/master/voxblox_ros/launch/euroc_dataset_bag.launch#L62

File: src/library/interface_mono.cpp [hunk 7]
```diff
@@ -32,18 +58,28 @@ void OrbSlam2InterfaceMono::imageCallback(const sensor_msgs::ImageConstPtr& msg)
...
     // Saving the transform to the member for publishing as a TF
-    T_W_C_ = T_W_C;
+    if(use_body_transform_)
+    {
+      T_W_B_ = T_B_C_*T_C_W.inverse();
+    }
+    else
+    {
+     T_W_B_ = T_C_W.inverse();
+    }
```
Comment (alexmillane) [inline src/library/interface_mono.cpp]: Formatting

File: src/library/interface_stereo.cpp [hunk 8]
```diff
@@ -6,38 +6,131 @@ namespace orb_slam_2_interface {
 
 OrbSlam2InterfaceStereo::OrbSlam2InterfaceStereo(const ros::NodeHandle& nh,
                                              const ros::NodeHandle& nh_private)
-    : OrbSlam2Interface(nh, nh_private) {
+    : OrbSlam2Interface(nh, nh_private),stereo_rectified_(false) {
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Formatting

File: src/library/interface_stereo.cpp [hunk 9]
```diff
@@ -6,38 +6,131 @@ namespace orb_slam_2_interface {
...
   // Getting data and params
   subscribeToTopics();
-  //advertiseTopics();
-  //getParametersFromRos();
+
+  if (!stereoRectification()){
+    ros::shutdown();
+    exit(1);
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Add an error message if you want to shutdown.

File: src/library/interface_stereo.cpp [hunk 10]
```diff
@@ -6,38 +6,131 @@ namespace orb_slam_2_interface {
...
 void OrbSlam2InterfaceStereo::subscribeToTopics() {
   // Subscribing to the stereo images
   left_sub_ = std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>>(
       new message_filters::Subscriber<sensor_msgs::Image>(
-          nh_, "camera/left/image_raw", 1));
+          nh_, "cam0/image_raw", 1));
   right_sub_ = std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>>(
       new message_filters::Subscriber<sensor_msgs::Image>(
-          nh_, "camera/right/image_raw", 1));
+          nh_, "cam1/image_raw", 1));
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Try not to change the default topic names to make them less readable. Please remap your topics in the launchfile to what you would like them to be.

http://wiki.ros.org/roslaunch/XML/remap

File: src/library/interface_stereo.cpp [hunk 11]
```diff
@@ -6,38 +6,131 @@ namespace orb_slam_2_interface {
...
   // Registering the synchronized image callback
   sync_->registerCallback(
       boost::bind(&OrbSlam2InterfaceStereo::stereoImageCallback, this, _1, _2));
+
+}
+
+bool OrbSlam2InterfaceStereo::getBodyTransform()
+{
+
+  cv::FileStorage fsSettings(settings_file_path_, cv::FileStorage::READ);
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: As above, use ros param server.

File: src/library/interface_stereo.cpp [hunk 12]
```diff
@@ -51,21 +144,41 @@ void OrbSlam2InterfaceStereo::stereoImageCallback(
...
+    cv::remap(cv_ptr_right->image,imRight,M1r_,M2r_,cv::INTER_LINEAR);
+    // Handing the image to ORB slam for tracking
+    T_C_W_opencv =
+        slam_system_->TrackStereo(imLeft, imRight,cv_ptr_left->header.stamp.toSec());
+  }
+  else
+  {
+    T_C_W_opencv =
+      slam_system_->TrackStereo(cv_ptr_left->image, cv_ptr_right->image,cv_ptr_left->header.stamp.toSec()); 
+  }
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Formatting

File: src/library/interface_stereo.cpp [hunk 13]
```diff
@@ -6,38 +6,131 @@ namespace orb_slam_2_interface {
...
+    cout << "LEFT.R" << R_l << endl;
+    cout << "RIGHT.R" << R_r << endl;
+    cout << "LEFT.P" << P_l << endl;
+    cout << "RIGHT.P" << P_r << endl;
+    cout << "Baseline" << -1*P_r.at<float>(0,3) << endl; // baseline
+    */
+
+    stereo_rectified_ = true;
+
+    return true;
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: We have a default node for stereo rectification here:

https://github.com/ethz-asl/image_undistort

if you find that you need more functionality. That said, this seems fairly light weight so it's fine to keep it in, with the obvious exception that params should come from the param server

File: src/library/interface_stereo.cpp [hunk 14]
```diff
@@ -51,21 +144,41 @@ void OrbSlam2InterfaceStereo::stereoImageCallback(
...
   }
-  // Handing the image to ORB slam for tracking
-  cv::Mat T_C_W_opencv =
-      slam_system_->TrackStereo(cv_ptr_left->image, cv_ptr_right->image,
-                                cv_ptr_left->header.stamp.toSec());
+
+  if(stereo_rectified_){
+    cv::Mat imLeft, imRight;
+    cv::remap(cv_ptr_left->image,imLeft,M1l_,M2l_,cv::INTER_LINEAR);
+    cv::remap(cv_ptr_right->image,imRight,M1r_,M2r_,cv::INTER_LINEAR);
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Cool. Nice. Cheers.

File: src/library/interface_stereo.cpp [hunk 15]
```diff
@@ -51,21 +144,41 @@ void OrbSlam2InterfaceStereo::stereoImageCallback(
...
     // Saving the transform to the member for publishing as a TF
-    T_W_C_ = T_W_C;
+    if(use_body_transform_)
+    {
+      T_W_B_ = T_B_C_*T_C_W.inverse();
+    }
+    else
+    {
+     T_W_B_ = T_C_W.inverse();
+    }
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Formatting as above

File: src/orb_slam_2_ros_node.cpp [hunk 16]
```diff
@@ -32,22 +34,26 @@ std::unique_ptr<orb_slam_2_interface::OrbSlam2Interface> create_interface(
...
   static const std::string kDefaultInterfaceType = "mono";
   std::string interface_type = kDefaultInterfaceType;
   nh_private.getParam("interface_type", interface_type);
+  
   // Creating the interface object to do the work
   std::unique_ptr<orb_slam_2_interface::OrbSlam2Interface> interface =
       create_interface(interface_type, nh, nh_private);
-  // Spinning
   ros::spin();
+  // Spinning
```
Comment (alexmillane) [inline src/orb_slam_2_ros_node.cpp]: ?

**Review by alexmillane — COMMENTED** [review]
See my comments. There's some stuff in here that can be easily fixed. My main comment is formatting and that we load params from the ros param server. Other than that, thank you for adding the option of rectification internal to the node.

**Comment by russellaabuchanan** [comment]
Hi Alex, thanks for your comments. I think I've fixed everything now, let me know what you think.

File: include/orb_slam_2_ros/interface_stereo.hpp [hunk 17]
```diff
@@ -36,6 +41,13 @@ class OrbSlam2InterfaceStereo : public OrbSlam2Interface {
   std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>> left_sub_;
   std::shared_ptr<message_filters::Subscriber<sensor_msgs::Image>> right_sub_;
   std::shared_ptr<message_filters::Synchronizer<sync_pol>> sync_;
+
+  // Rectification maps
+  cv::Mat M1l_, M2l_, M1r_, M2r_;
```
Comment (alexmillane) [inline include/orb_slam_2_ros/interface_stereo.hpp]: Thanks for the comment :+1:. But on second thought, can you rename these descriptively:

https://google.github.io/styleguide/cppguide.html#General_Naming_Rules

File: launch/run_orb_slam_2.launch [hunk 18]
```diff
@@ -3,10 +3,17 @@
...
-    <remap from="camera/right/image_raw" to="/cam01/image_raw" />
+    <param name="vocabulary_file_path" value="/home/raab/ros_vios/catkin_ws/build/orb_slam_2_catkin/orb_slam_2_src-prefix/src/orb_slam_2_src/Vocabulary/ORBvoc.txt" />
+    <param name="settings_file_path" value="/home/raab/ros_vios/catkin_ws/build/orb_slam_2_catkin/orb_slam_2_src-prefix/src/orb_slam_2_src/Settings/vios_calibration.yaml" />
+    <param name="visualization" type="bool" value="true" /> <!-- Disale when running on quadcotper -->
+    <param name="use_body_transform" type="bool" value="false" /> <!-- Body transform isn't available in vanilla Orb Slam 2 -->
+
+    <!-- param name="vocabulary_file_path" value="/home/millanea/trunk/datasets/kitti/raw/ORBvoc.txt" /-->
+    <!-- param name="settings_file_path" value="/home/millanea/trunk/datasets/kitti/raw/2011_09_30_drive_18/KITTI00-02_stereo.yaml" /-->
+    <remap from="camera/left/image_raw" to="/cam0/image_raw" />
+	<remap from="camera/right/image_raw" to="/cam1/image_raw" />
```
Comment (alexmillane) [inline launch/run_orb_slam_2.launch]: Good work :+1:

File: src/library/interface.cpp [hunk 19]
```diff
@@ -14,10 +14,18 @@ OrbSlam2Interface::OrbSlam2Interface(const ros::NodeHandle& nh,
       nh_private_(nh_private),
       verbose_(kDefaultVerbose),
       frame_id_(kDefaultFrameId),
-      child_frame_id_(kDefaultChildFrameId) {
-  // Getting data and params
+      child_frame_id_(kDefaultChildFrameId),
+      visualization_(kDefaultVisualization),
+      use_body_transform_(kDefaultUseBodyTransform){
```
Comment (alexmillane) [inline src/library/interface.cpp]: Format

File: src/library/interface.cpp [hunk 20]
```diff
@@ -14,10 +14,18 @@ OrbSlam2Interface::OrbSlam2Interface(const ros::NodeHandle& nh,
...
+
   advertiseTopics();
   getParametersFromRos();
+
+  if(use_body_transform_)
```
Comment (alexmillane) [inline src/library/interface.cpp]: Auto-format this

File: src/library/interface.cpp [hunk 21]
```diff
@@ -83,4 +94,24 @@ void OrbSlam2Interface::convertOrbSlamPoseToKindr(const cv::Mat& T_cv,
   *T_kindr = Transformation(q_kindr, t_kindr);
 }
 
+void OrbSlam2Interface::getBodyTransform() {
+  cv::FileStorage fsSettings(settings_file_path_, cv::FileStorage::READ);
```
Comment (alexmillane) [inline src/library/interface.cpp]: Please load settings using rosparams. I can show you how to do this if you like.

File: src/library/interface_stereo.cpp [hunk 22]
```diff
@@ -33,11 +35,69 @@ void OrbSlam2InterfaceStereo::subscribeToTopics() {
       boost::bind(&OrbSlam2InterfaceStereo::stereoImageCallback, this, _1, _2));
 }
 
+void OrbSlam2InterfaceStereo::stereoRectification() {
+  // Load settings related to stereo calibration
+  cv::FileStorage fsSettings(settings_file_path_, cv::FileStorage::READ);
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: Again please load settings using rosparam.

File: src/library/interface_stereo.cpp [hunk 23]
```diff
@@ -33,11 +35,69 @@ void OrbSlam2InterfaceStereo::subscribeToTopics() {
...
+  cv::stereoRectify(K_l, D_l, K_r, D_r, cv::Size(cols_l, rows_l),
+                    T_C0_1C.rowRange(0, 3).colRange(0, 3),
+                    T_C0_1C.col(3).rowRange(0, 3), R_l, R_r, P_l, P_r, Q_);
+
+  cv::initUndistortRectifyMap(K_l, D_l, R_l, P_l.rowRange(0, 3).colRange(0, 3),
+                              cv::Size(cols_l, rows_l), CV_32F, M1l_, M2l_);
+  cv::initUndistortRectifyMap(K_r, D_r, R_r, P_r.rowRange(0, 3).colRange(0, 3),
+                              cv::Size(cols_r, rows_r), CV_32F, M1r_, M2r_);
+
+  // Uncomment if you want to see the rectified instrinsics ex. to put in yaml
```
Comment (alexmillane) [inline src/library/interface_stereo.cpp]: In general code which you have to uncomment to use shouldn't be put into the master branch. Either expose the option to output these to the command line with a rosparam flag, or remove entirely. Of course this is often useful in development, so I encourage you to use it on a development branch.

**Review by alexmillane — CHANGES_REQUESTED** [review]
Thank you for your changes. Mostly fixed. I think your autoformatter missed a few things, and again, please use the rosparameter server to load parameters.

**Comment by russellaabuchanan** [comment]
I fixed my auto formatter so hopefully that's all good now. I added a check for projection and rectification matrices since although they can be calculated easily they're actually included in the default ORB_SLAM2 yaml parameter files. Also I left in getting the parameters from the yamls as per our discussion.
